File: ontrack-extension-scm/src/main/java/net/nemerosa/ontrack/extension/scm/changelog/SCMChangeLog.java
package net.nemerosa.ontrack.extension.scm.changelog;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.Data;
import lombok.EqualsAndHashCode;
import net.nemerosa.ontrack.model.structure.Branch;
import net.nemerosa.ontrack.model.buildfilter.BuildDiff;
import net.nemerosa.ontrack.model.structure.BuildView;

import java.util.UUID;

/**
 * @param <S> Type of SCM data associated with the branch
 * @param <T> Type of SCM data associated with a build
 */
@EqualsAndHashCode(callSuper = false)
@Data
public class SCMChangeLog<S, T> extends BuildDiff {

    private final String uuid;
    @JsonIgnore
    private final S scmBranch;
    private final SCMBuildView<T> scmBuildFrom;
    private final SCMBuildView<T> scmBuildTo;

    protected SCMChangeLog(String uuid, Branch branch, S scmBranch, SCMBuildView<T> scmBuildFrom, SCMBuildView<T> scmBuildTo) {
        super(branch);
        this.uuid = uuid;
        this.scmBranch = scmBranch;
        this.scmBuildFrom = scmBuildFrom;
        this.scmBuildTo = scmBuildTo;
    }

    @Override
    public BuildView getFrom() {
        return scmBuildFrom.getBuildView();
    }

    @Override
    public BuildView getTo() {
        return scmBuildTo.getBuildView();
    }

    public static <S, T> SCMChangeLog<S, T> of(Branch branch, S scmBranch, SCMBuildView<T> from, SCMBuildView<T> to) {
        // TODO Checks the order of the builds
        return new SCMChangeLog<>(
                UUID.randomUUID().toString(),
                branch,
                scmBranch,
                from,
                to
        );
    }

}

File: ontrack-extension-scm/src/main/java/net/nemerosa/ontrack/extension/scm/changelog/SCMChangeLog.kt
package net.nemerosa.ontrack.extension.scm.changelog

import net.nemerosa.ontrack.extension.scm.model.SCMChangeLogCommit
import net.nemerosa.ontrack.model.annotations.APIDescription
import net.nemerosa.ontrack.model.structure.Build

@APIDescription("Representation of a change log between two builds")
data class SCMChangeLog(
    @APIDescription("Boundary for the change log")
    val from: Build,
    @APIDescription("Boundary for the change log")
    val to: Build,
    @APIDescription("Boundary commit for the change log")
    val fromCommit: String,
    @APIDescription("Boundary commit for the change log")
    val toCommit: String,
    @APIDescription("List of commits between the builds")
    val commits: List<SCMDecoratedCommit>,
    @APIDescription("List of issues between the builds")
    val issues: SCMChangeLogIssues?,
) {
    fun isEmpty() = fromCommit == toCommit
}
